File: src/main/java/com/werfad/KeyTagsGenerator.java
package com.werfad;

import com.werfad.utils.ArrayUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class KeyTagsGenerator {
    // maybe has a math method
    public static List<String> createTagsTree(int targetCount, String keys) {
        List<String> res = new ArrayList<>();
        _createTagsTree(targetCount, keys, "", res);
        return res;
    }

    private static void _createTagsTree(int targetCount, String keys, String prefix, List<String> res) {
        int keysLen = keys.length();
        HashMap<Character, Integer> keysCountMap = new HashMap<>();

        char[] chars = keys.toCharArray();
        for (char c : chars) {
            keysCountMap.put(c, 0);
        }

        /* Calculate counts each branch stored. */
        ArrayUtils.revertChars(chars);
        int targetsLeft = targetCount;
        int level = 0;

        while (targetsLeft > 0) {
            int childsLen = (level == 0 ? 1 : (keysLen - 1));

            for (char key : chars) {
                keysCountMap.put(key, keysCountMap.get(key) + childsLen);

                targetsLeft -= childsLen;
                if (targetsLeft <= 0) {
                    keysCountMap.put(key, keysCountMap.get(key) + targetsLeft);
                    break;
                }
            }
            level++;
        }

        /* Create tree ( represent by String array ). */
        ArrayUtils.revertChars(chars);
        for (char key : chars) {
            int count = keysCountMap.get(key);
            if (count > 1) {
                _createTagsTree(count, keys, prefix + key, res);
            } else if (count == 1) {
                res.add(prefix + key);
            }
        }
    }
}

File: src/main/kotlin/com/werfad/KeyTagsGenerator.kt
package com.werfad

object KeyTagsGenerator {
    // maybe has a math method
    fun createTagsTree(targetCount: Int, keys: String): List<String> {
        val res: MutableList<String> = ArrayList()
        recur(targetCount, keys, "", res)
        return res
    }

    private fun recur(targetCount: Int, keys: String, prefix: String, res: MutableList<String>) {
        val keysLen = keys.length
        val keysCountMap = HashMap<Char, Int>()
        val chars = keys.toCharArray()
        for (c in chars) {
            keysCountMap[c] = 0
        }

        /* Calculate counts each branch stored. */
        chars.reverse()
        var targetsLeft = targetCount
        var level = 0
        while (targetsLeft > 0) {
            val childLen = if (level == 0) 1 else keysLen - 1
            for (key in chars) {
                keysCountMap[key] = keysCountMap[key]!! + childLen
                targetsLeft -= childLen
                if (targetsLeft <= 0) {
                    keysCountMap[key] = keysCountMap[key]!! + targetsLeft
                    break
                }
            }
            level++
        }

        /* Create tree ( represent by String array ). */
        chars.reverse()
        for (key in chars) {
            val count = keysCountMap[key]!!
            if (count > 1) {
                recur(count, keys, prefix + key, res)
            } else if (count == 1) {
                res.add(prefix + key)
            }
        }
    }
}